core/envelope-sim/src/test/java/fr/sncf/osrd/envelope/EnvelopeTest.kt

```diff
@@ -0,0 +1,204 @@
+    }
+
+    @Test
+    fun testInterpolateDepartureFrom() {
+        val partA =
+            EnvelopeTestUtils.generateTimes(doubleArrayOf(0.0, 2.0), doubleArrayOf(1.0, 1.0))
+        val partB =
+            EnvelopeTestUtils.generateTimes(doubleArrayOf(2.0, 4.0), doubleArrayOf(1.0, 1.0))
+        val envelope = make(partA, partB)
+
+        assertEquals(1.0, partA.interpolateTotalTime(1.0))
+
+        assertEquals(1.0, envelope.interpolateDepartureFrom(1.0))
+        assertEquals(2.0, envelope.interpolateDepartureFrom(2.0))
+        assertEquals(2.0, envelope.interpolateDepartureFrom(2.0))
+        assertEquals(3.0, envelope.interpolateDepartureFrom(3.0))
+        assertEquals(3.5, envelope.interpolateDepartureFrom(3.5))
+        assertEquals(4.0, envelope.interpolateDepartureFrom(4.0))
+    }
+
+    @Test
+    fun testMinEnvelopesThrowsWhenDifferentPositionRanges() {
+        val envelopeA =
+            make(
+                EnvelopeTestUtils.generateTimes(
+                    doubleArrayOf(0.0, 10.0),
+                    doubleArrayOf(100.0, 100.0)
+                )
+            )
+        val envelopeB =
+            make(
+                EnvelopeTestUtils.generateTimes(
+                    doubleArrayOf(5.0, 10.0),
+                    doubleArrayOf(100.0, 100.0)
+                )
+            )
+        val envelopeC =
+            make(
+                EnvelopeTestUtils.generateTimes(
+                    doubleArrayOf(0.0, 15.0),
+                    doubleArrayOf(100.0, 100.0)
+                )
+            )
+        val envelopeD =
+            make(
+                EnvelopeTestUtils.generateTimes(
+                    doubleArrayOf(5.0, 15.0),
+                    doubleArrayOf(100.0, 100.0)
+                )
+            )
+        assertThrows(AssertionError::class.java) { minEnvelopes(envelopeA, envelopeB) }
+        assertThrows(AssertionError::class.java) { minEnvelopes(envelopeA, envelopeC) }
+        assertThrows(AssertionError::class.java) { minEnvelopes(envelopeA, envelopeD) }
+    }
+
+    @ParameterizedTest
+    @MethodSource("minEnvelopePartsArgs")
+    fun testMinEnvelopeParts(
+        envelopeA: Envelope,
+        envelopeB: Envelope,
+        intersectionIndex: Int,
+        expectedPositions: List<Double>,
+        expectedSpeeds: List<Double>,
+    ) {
+        val minEnvelope = minEnvelopes(envelopeA, envelopeB)
+        val resultingPoints = minEnvelope.iteratePoints()
+        val resultingPositions = resultingPoints.map { it.position }
+        val resultingSpeeds = resultingPoints.map { it.speed }
+
+        // Check that both envelopes have an equal speed at their intersection.
+        val intersectionPosition = resultingPositions[intersectionIndex]
+        val intersectionSpeed = resultingSpeeds[intersectionIndex]
+        assertTrue(
+            areSpeedsEqual(
+                envelopeA.interpolateSpeedLeftDir(intersectionPosition, 1.0),
```

Comment:
nit: as I wondered what the 1.0 was (same below)
```suggestion
        directionAscending = 1.0
        assertTrue(
            areSpeedsEqual(
                envelopeA.interpolateSpeedLeftDir(intersectionPosition, directionAscending),
```